pkg/operator/controller/gateway-labeler/controller.go

```diff
@@ -0,0 +1,222 @@
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+const (
+	controllerName = "gateway_labeler_controller"
+)
+
+var log = logf.Logger.WithName(controllerName)
+
+// NewUnmanaged creates and returns a controller that adds the "istio.io/rev"
+// label to gateways so that Istio reconciles them.  This is an unmanaged
+// controller, which means that the manager does not start it.
+//
+// The "istio.io/rev" label MUST be present on gateways in order for an
+// Istio control-plane deployed via the OSSM Operator to consider that
+// resource managed.
+func NewUnmanaged(mgr manager.Manager) (controller.Controller, error) {
+	// Create a new cache for gateways so it can watch all namespaces.
+	// (Using the operator cache for gateways in all namespaces would cause
+	// it to cache other resources in all namespaces.)
+	gatewaysCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Make sure the manager starts the cache with the other runnables.
+	mgr.Add(gatewaysCache)
+
+	reconciler := &reconciler{
+		cache:    gatewaysCache,
+		client:   mgr.GetClient(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	// Watch gatewayclasses so that when one is created with our controller
+	// name or is updated to change its controller name to our controller
+	// name, then we update labels on gateways that reference that
+	// gatewayclass.
+	key, value := operatorcontroller.IstioRevLabelKey, operatorcontroller.IstioName("").Name
+	gatewayClassToGateways := func(ctx context.Context, o client.Object) []reconcile.Request {
+		requests := []reconcile.Request{}
+
+		// Using an index might be more efficient than listing all
+		// gateways and then filtering.  However, setting up an index is
+		// complicated by the fact that the Gateway API CRDs might not
+		// exist when this controller is initialized.
+
+		var gateways gatewayapiv1.GatewayList
+		if err := reconciler.cache.List(ctx, &gateways); err != nil {
+			log.Error(err, "Failed to list gateways for gatewayclass", "gatewayclass", o.GetName())
+			return requests
+		}
+
+		for i := range gateways.Items {
+			if gateways.Items[i].Labels[key] == value {
+				continue
+			}
+			if string(gateways.Items[i].Spec.GatewayClassName) != o.GetName() {
+				continue
+			}
+			request := reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: gateways.Items[i].Namespace,
+					Name:      gateways.Items[i].Name,
+				},
+			}
+			requests = append(requests, request)
+		}
```

Comment:
nit:

```suggestion
		for _, gateway := range gateways.Items {
			if gateway.Labels[key] == value {
				continue
			}
			if string(gateway.Spec.GatewayClassName) != o.GetName() {
				continue
			}
			request := reconcile.Request{
				NamespacedName: types.NamespacedName{
					Namespace: gateway.Namespace,
					Name:      gateway.Name,
				},
			}
			requests = append(requests, request)
		}
```

Reply:
We've been tending towards using the index to avoid copying large values in the case of struct types, especially in hot loops or with large slices.  For example, using the index can yield a significant performance improvement when iterating over thousands of routes.  For the code in question, this might not be a particularly hot loop, and there probably aren't thousands or even hundreds of gateways, but I still prefer not to copy the gateway if we only need a few of its fields.